1832 add tutorial for temporal encoding (#1834)
informalsystems/apalache: docs/src/tutorials/temporal-properties.md

and a button which can be pushed to request the traffic light to become green.
Consequently, there are two variables:
the current state of the light (either green or red),
and whether the button has been pushed that requests the traffic light to switch from red to green.

The full specification of the traffic light is here:
[TrafficLight.tla](TrafficLight.tla)
But don't worry - we will dissect the spec in the following.

In the TLA specification, this corresponds to
two variables:

```
{{#include TrafficLight.tla:2:15}}
```

Initially, the light should be red and green should not be requested:

```
{{#include TrafficLight.tla:18:20}}
```

We have three possible actions: 
1. The traffic light can switch from red to green, 
2. The traffic light can switch from green to red, or
3. The button can be pushed, thus requesting that the traffic light becomes green.

```
{{#include TrafficLight.tla:actions}}
```

Now, we are ready to specify properties we are interested in.
For example, when green is requested, at some point afterwards the light should actually turn green.
We can write the property like this:

```
{{#include TrafficLight.tla:prop}}

> ```

Author: Added this below

(* State3 ==
    ...
    /\ __loop_InLoop = TRUE
    ...
    /\ __loop_isGreen = FALSE
    /\ __loop_requestedGreen = TRUE
    ...
    /\ isGreen = FALSE
    /\ requestedGreen = TRUE *)
```
So, initially, `isGreen` and `requestedGreen` are both false.
Further, `__loop_InLoop` is false, and the copies of `isGreen` and `requestedGreen`, which are called
`__loop_isGreen` and `__loop_requestedGreen`, are equal to the values of `isGreen` and `requestedGreen`.

From state 0 to state 1, `requestedGreen` changes from false to true.
From state 1 to state 2, the system stutters, and nothing seems to change (in fact,
some auxiliary variables 'under the hood' change, which we will dive into later).
Finally, in state 3 `__loop_InLoop` is set to true, which means that
the loop starts in state 2, and the trace from state 3 onward is inside the loop.
However, since state 3 is the last state, this means simply that the trace loops in state 2.
Since the loop starts, the copies of the system variables are also set to the values of the variables in state 2,
so ` __loop_isGreen = FALSE` and `__loop_requestedGreen = TRUE`.

The looping trace in this case can be visualized like this:

> ![A counterexample trace for the property [](requestedGreen => <>isGreen)](img/counterexample.png)

Author: Should be there now

It is also clear why this trace violates the property:
`requestedGreen` holds in state 1, but `isGreen` never holds,
so in state 1 the property `requestedGreen => <>isGreen` is violated.

## Auxiliary variables encode evaluations of subformulas along the trace

Next, let us discuss the other auxiliary variables that are introduced by Apalache to check the temporal property.
These extra variables correspond to parts of the temporal property we want to check.
These are the following variables with their valuations in state 0:

```
(* State0 ==
  RequestWillBeFulfilled_init = FALSE
    ...
    /\ ☐(requestedGreen ⇒ ♢isGreen) = FALSE
    /\ ☐(requestedGreen ⇒ ♢isGreen)_unroll = TRUE
    /\ requestedGreen ⇒ ♢isGreen = TRUE
    /\ ♢isGreen = FALSE
    /\ ♢isGreen_unroll = FALSE
    ...
```
There are three groups of variables: those that look like formulas, e.g. `☐(requestedGreen ⇒ ♢isGreen)`,
those that look like formulas and end with `_unroll`,

> e.g., and the variable `RequestWillBeFulfilled_init`.

Reviewer: This is not resolved yet.

us that along the execution, `isGreen` will never be true.

You might already have noticed the pattern of which formulas appear as variables.
Take our property `[](requestedGreen => <>isGreen)`.
The syntax tree of this formula looks like this:

![The syntax tree for the formula [](requestedGreen => <>isGreen)](img/syntaxtree.png)

For each node of the syntax tree where the formula contains a temporal operator,
there is an auxiliary variable. And as mentioned before, the value of 
the variable in a state tells us whether from that state, the subformula is true. In this particular example, the formulas that correspond to variables in the encoding are filled with orange in the syntax tree.

What about the `_unroll` variables? There is one `_unroll` variable for each temporal operator in the formula.
For example, `☐(requestedGreen ⇒ ♢isGreen)_unroll` is the unroll-variable for the
leading box operator. These are necessary because to know whether a formula like
`[]isGreen` holds in the last state of the loop, we need to know whether
`isGreen` holds in all states of the loop. However, when we are in the last state of the loop,

> we won't see the state of the loop again! So we need to store this information when we traverse the loop.

Reviewer: I don't understand what state "we won't see the state of the loop again".

Also, if something is done by an algorithm, I prefer to say it that we instead of using a generic "we".

> we won't see the state of the loop again! So we need to store this information when we traverse the loop.

Author: Rewrote this a bit and replaced "we" where appropriate

> That's why we have an extra variable, which essentially tells us whether `isGreen` holds on all states of the loop.

Author: I added how this works for `<>` operators

    ...
```

So in the last state, `☐(requestedGreen ⇒ ♢isGreen)_unroll`
is not true, since `☐(requestedGreen ⇒ ♢isGreen)`
does not hold in state 2, which is on the loop.

Notice that for the variables corresponding to nodes in the
syntax tree, e.g. `☐(requestedGreen ⇒ ♢isGreen)`,
we have loop copies, e.g. `__loop_☐(requestedGreen ⇒ ♢isGreen)`.

These fulfill the same function as the `__loop_` copies for the
original variables of the model, as explained above.

```
(* State0 ==
    ...
    /\ __loop_☐(requestedGreen ⇒ ♢isGreen) = FALSE
    /\ __loop_requestedGreen ⇒ ♢isGreen = TRUE
    /\ __loop_♢isGreen = FALSE
    /\ __loop_isGreen = FALSE
    /\ __loop_requestedGreen = FALSE
```

Finally, let's discuss `RequestWillBeFulfilled_init`.
Recall that the variables for subformulas store
the evaluations of the subformulas at the current point in the execution.
This means that when we are at the end of the execution, we don't know
whether the variable for the whole formula, e.g. `☐(requestedGreen ⇒ ♢isGreen)`, was true in the first state.
That's what `RequestWillBeFulfilled_init` is for:
it tells us whether in the first state of the execution, our temporal
property holds true

> (recall that we named the property "RequestWillBeFulfilled").

Author: I added more explanation.

> (recall that we named the property "RequestWillBeFulfilled").

Reviewer: I'm afraid I still don't understand.  This is probably obvious if one knows the entire construction, which I don't.
To clarify, is `☐(requestedGreen ⇒ ♢isGreen)` in the first state a prophecy variable, or is it constrained st we get a model value for it after finding a counter-example / exhaustively exploring the search space?

> (recall that we named the property "RequestWillBeFulfilled").

Author: It is a "prohecy variable". If the variable is false, then
in this trace, the formula will not hold. However, only a trace where the prohecy variable says false, and the model checker finds a valid loop (satisfying some constraints wrt the model and aux. variables) is a problem, so the prophecy variable being false does not yet constitute a bad trace. It being false and then later being able to find a valid loop is a counterexample.

> (recall that we named the property "RequestWillBeFulfilled").

Reviewer: I see, thanks for the explanation!
It seems to me that there is no way of motivating the `_init` variable without knowing the construction. Which would be useful for someone to understand the construction, but this tutorial is aimed at users checking temporal properties. Maybe we should just say that it is an artifact of the model checking algorithm?

> (recall that we named the property "RequestWillBeFulfilled").

Reviewer: We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

> (recall that we named the property "RequestWillBeFulfilled").

Reviewer: I think this is still an issue.

WDYT about my suggestion above?
> Maybe we should just say that it is an artifact of the model checking algorithm?
>  We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

> (recall that we named the property "RequestWillBeFulfilled").

Author: My bad, one commit got lost on the way. I've added a paragraph.

> (recall that we named the property "RequestWillBeFulfilled").

Reviewer: looks good, thx!

A loop is a partial trace that starts and ends with the same state. 
A trace containing a loop describes a possible infinite execution that repeats the loop forever.

For example, what is a trace that is a counterexample to the property `♢isGreen`?
It's an execution that loops without ever finding a state that satisfies `isGreen`.
For example, a counterexample trace might visually look like this:

![A counterexample trace for the property <>isGreen](img/looping_trace.png)

In contrast, as long as the model checking engine has not found a loop, there may still exist some future state satisfying `isGreen`.

## Utilizing auxiliary variables to find loops

The encoding for temporal properties involves lots of auxiliary variables.
While some can be very helpful to understand counterexamples,
many are mostly noise.

Let's first understand how Apalache can identify looping executions using auxiliary variables.
The auxiliary variable `__loop_InLoop` is true in exactly the states belonging to the loop.
Additionally, at the first state of the loop, i.e., when `__loop_InLoop` switches from false to true,

> we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.

Author: ```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
```

```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
Before the first state of the loop, the values of the `__loop_` variables is unspecified.
```

# Temporal properties and counterexamples

**Difficulty: Red trail – Medium**

> 

Reviewer 2: Add yourself as an author.

In this tutorial, we will show how Apalache can be used to decide temporal
properties that are more general than invariants.
This tutorial will be most useful to you if you have a basic understanding of
linear temporal logic, e.g. the semantics of `<>` and `[]` operators.

> Further, we assume you are familiar with TLA+, but expert knowledge is not necessary.

Reviewer 2: I wanted to give a link to a video lecture by Leslie Lamport on that topic, but it looks like there is no lecture where he is presenting `<>`, or maybe I have not found it. Maybe give a reference to "Specifying Systems"?

Reviewer: There is a brief explanation of `<>` in video 9a.
But maybe Hillel's page on temporal operators is more informative? https://learntla.com/temporal-logic/operators/

Author: Added a link to Hilel's page

    /\ __temporal_t_2 = TRUE
    /\ __temporal_t_3 = FALSE
    /\ __temporal_t_3_unroll = FALSE
    /\ isGreen = FALSE
    /\ requestedGreen = FALSE
```

Two things are notable:
1. The initial state formula appears twice, once as a comment and once in TLA+.
2. There are way more variables than the two variables we specified.

The comment and the TLA+ specification express the same state, but in the comment, some variable names from the encoding have been replaced
with more human-readable names.
For example, there is a variable called `☐(requestedGreen ⇒ ♢isGreen)` in the comment,
which is called `__temporal_t_1` in TLA+.
In the following, let's focus on the content of the comment, since it's easier to understand what's going on.

There are many additional variables in the counterexample because to check temporal formulas, Apalache uses an
encoding that transforms temporal properties to invariants.
If you are interested in the technical details, the encoding is described in sections 3.2 and 4 of [Biere et al.][].
However, to understand the counterexample, you don't need to go into the technical details of the encoding.
We'll go explain the counterexample in the following.

> ## Counterexamples encode traces with a loop

Reviewer 2: The notion of a trace appears out of nowhere. We usually use the word "computation" and the canonical TLA+ notion is that of a behavior. I think you are right in that we have to talk about traces here, as we decorate computation states with additional predicates. It would be good to have the notion of a trace quickly introduced in one paragraph.

Author: The nomenclature is from https://apalache.informal.systems/docs//tutorials/symbmc.html?highlight=trace#symbolic-traces, so I'll just link that

First, it's important to know that counterexamples to temporal properties are, in general, traces ending in a loop.
A loop is a partial trace that starts and ends with the same state. 
A trace containing a loop describes a possible infinite execution that repeats the loop forever.

For example, what is a trace that is a counterexample to the property `♢isGreen`?
It's an execution that loops without ever finding a state that satisfies `isGreen`.
For example, a counterexample trace might visually look like this:

![A counterexample trace for the property <>isGreen](img/looping_trace.png)

In contrast, as long as the model checking engine has not found a loop, there may still exist some future state satisfying `isGreen`.

## Utilizing auxiliary variables to find loops

The encoding for temporal properties involves lots of auxiliary variables.
While some can be very helpful to understand counterexamples,
many are mostly noise.

> Let's first understand how Apalache can identify looping executions using auxiliary variables.

Reviewer 2: Let's use the word "lasso" for executions with a loop, to have the stable vocabulary.

The auxiliary variable `__loop_InLoop` is true in exactly the states belonging to the loop.
Additionally, at the first state of the loop, i.e., when `__loop_InLoop` switches from false to true,

> we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.

Reviewer 2: ```suggestion
we store the valuation of all variables in a shadow copy whose name is prefixed by `__loop_`.
```

> Before the first state of the loop, the values of the `__loop_` variables are unspecified.

Reviewer 2: ```suggestion
Before the first state of the loop, the values of the `__loop_` variables may carry arbitrary values.
```

They are specified in the counterexample.

## Auxiliary variables encode evaluations of subformulas along the trace

Next, let us discuss the other auxiliary variables that are introduced by Apalache to check the temporal property.
These extra variables correspond to parts of the temporal property we want to check.
These are the following variables with their valuations in the initial state:

```
(* State0 ==
  RequestWillBeFulfilled_init = FALSE
    ...
    /\ ☐(requestedGreen ⇒ ♢isGreen) = FALSE
    /\ ☐(requestedGreen ⇒ ♢isGreen)_unroll = TRUE
    /\ requestedGreen ⇒ ♢isGreen = TRUE
    /\ ♢isGreen = FALSE
    /\ ♢isGreen_unroll = FALSE
    ...
```
There are three groups of variables: 
* Variables that look like formulas, e.g. `☐(requestedGreen ⇒ ♢isGreen)`
* Variables that look like formulas and end with `_unroll`,
e.g. `☐(requestedGreen ⇒ ♢isGreen)_unroll`
* The variable `RequestWillBeFulfilled_init`.

Let's focus on the non-`_unroll` variables that look like formulas
first.

Recall that the temporal property we want to check is `[](requestedGreen => <>isGreen)`.
That's also the name of one of the variables: The value of the variable 
`☐(requestedGreen ⇒ ♢isGreen)` tells us whether starting in the current state, the

> formula `[](requestedGreen => <>isGreen)` holds. Since we are looking at a counterexample, it is not
> surprising that the formula does not hold in state 0.

Reviewer 2: ```suggestion
formula `[](requestedGreen => <>isGreen)` holds. Since we are looking at a counterexample to this formula, it should not be
surprising that the formula does not hold in the initial state of the counterexample.
```
